File: 329/solution.rb
# DFS + memorization

# @param {Integer[][]} matrix
# @return {Integer}
def longest_increasing_path(matrix)
    result = 0
    return result if matrix.empty?
    n, m = matrix.size, matrix[0].size
    dp = Array.new(n) {Array.new(m, -1)}
    
    n.times do |i|
        m.times do |j|
            result = [result, DFS(matrix, i, j, n, m, dp)].max
        end
    end
    
    result
end

def DFS(matrix, i, j, n, m, dp)
    return dp[i][j] if dp[i][j] != -1
    
    dp[i][j] = 1
    dirs = [0, 1, 0, -1, 0]
    4.times do |k|
        x, y = i + dirs[k], j + dirs[k + 1]
        if x < 0 || y < 0 || x >= n || y >= m || matrix[x][y] <= matrix[i][j]
            next
        end
        dp[i][j] = [dp[i][j], 1 + DFS(matrix, x, y, n, m, dp)].max
    end
    
    dp[i][j]
end

File: 329/solution.c
// DFS + memorization

inline int max(const int a, const int b) {
    return a > b ? a : b;
}

int DFS(int** matrix, int i, int j, int n, int m, int** dp);

int longestIncreasingPath(int** matrix, int matrixSize, int* matrixColSize){
    int result = 0;
    if(matrixSize <= 0 || *matrixColSize <= 0) {
        return result;
    }
    int n = matrixSize, m = *matrixColSize;
    int** dp;
    dp = (int **)malloc(sizeof(int *) * n);
    for(int i = 0; i < n; i++) {
        dp[i] = (int *)malloc(sizeof(int) * m);
    }
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
            dp[i][j] = -1;
        }
    }
    
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
            result = max(result, DFS(matrix, i, j, n, m, dp));
        }
    }
    
    return result;
}

int DFS(int** matrix, int i, int j, int n, int m, int** dp) {
    if(dp[i][j] != -1) {
        return dp[i][j];
    }
    
    static int dirs[] = {0, 1, 0, -1, 0};
    dp[i][j] = 1;
    for(int k = 0; k < 4; k++) {
        int x = i + dirs[k], y = j + dirs[k + 1];
        if(x < 0 || y < 0 || x >= n || y >= m || matrix[x][y] <= matrix[i][j]) {
            continue;
        }
        dp[i][j] = max(dp[i][j], 1 + DFS(matrix, x, y, n, m, dp));
    }
    
    return dp[i][j];
}


